pkg/controller/backup_controller.go

```diff
@@ -666,17 +799,106 @@ func (c *backupController) runBackup(backup *pkgbackup.Request) error {
 		return err
 	}
 
-	if errs := persistBackup(backup, backupFile, logFile, backupStore, c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)), volumeSnapshots, volumeSnapshotContents); len(errs) > 0 {
+	if errs := persistBackupBeforeUploadComplete(backup, backupFile, logFile, backupStore, volumeSnapshots, volumeSnapshotContents); len(errs) > 0 {
 		fatalErrs = append(fatalErrs, errs...)
 	}
 
-	c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup completed")
+	snapshotInProgress := false
+	if features.IsEnabled(velerov1api.UploadProgressFeatureFlag) {
+		snapshotInProgress = checkAndUpdateItemSnapshotsProgress(backup.ItemSnapshots, logger)
+	}
+
+	if snapshotInProgress {
+		// Prepare for upload progress monitoring
+		switch {
+		case len(fatalErrs) > 0:
+			// If we had fatal errors in the backup, we will just fail here and not monitor upload progress
+			// ItemSnapshotsAttempted/Completed may be incorrect
+			backup.Status.Phase = velerov1api.BackupPhaseFailed
+		case logCounter.GetCount(logrus.ErrorLevel) > 0:
+			backup.Status.Phase = velerov1api.BackupPhaseUploadingPartialFailure
+			c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Backup partially failed but snapshot uploads outstanding, will complete in background")
+		default:
+			backup.Status.Phase = velerov1api.BackupPhaseUploading
+			c.logger.WithField(Backup, kubeutil.NamespaceAndName(backup)).Info("Snapshot uploads outstanding, will complete in background")
+		}
+	} else {
+		// Snapshot uploads either didn't happen or they are complete so update ItemSnapshotAttempted/Completed and
+		// go to our final phases
+		backup.Status.ItemSnapshotsAttempted = len(backup.ItemSnapshots)
+		for _, snap := range backup.ItemSnapshots {
+			if snap.Status.Phase == isv1.SnapshotPhaseCompleted {
+				backup.Status.ItemSnapshotsCompleted++
+			}
+		}
+
+		switch {
+		case len(fatalErrs) > 0:
+			backup.Status.Phase = velerov1api.BackupPhaseFailed
+		case logCounter.GetCount(logrus.ErrorLevel) > 0:
```

Comment:
Why not compare the `ItemSnapshotsAttempted` and `ItemSnapshotsCompleted`

Reply:
By the time we get to this section, all snapshots have moved into a final stage either SnapshotPhaseCompleted or SnapshotPhaseFailed and if they failed an error should have been logged.